service/history/api/startworkflow/api.go

```diff
@@ -41,167 +46,463 @@ import (
+		workflowConditionFailed.Status,
 		workflowID,
-		runID.String(),
-		startRequest.StartRequest.GetWorkflowIdReusePolicy(),
+		runID,
+		s.request.StartRequest.GetWorkflowIdReusePolicy(),
 	)
 	if err != nil {
 		return nil, err
 	}
 
-	if prevExecutionUpdateAction != nil {
-		// update prev execution and create new execution in one transaction
-		err := api.GetAndUpdateWorkflowWithNew(
-			ctx,
-			nil,
-			api.BypassMutableStateConsistencyPredicate,
-			definition.NewWorkflowKey(
-				namespaceID.String(),
+	if prevExecutionUpdateAction == nil {
+		return nil, nil
+	}
+	var mutableStateInfo *mutableStateInfo
+	// update prev execution and create new execution in one transaction
+	err = api.GetAndUpdateWorkflowWithNew(
+		ctx,
+		nil,
+		api.BypassMutableStateConsistencyPredicate,
+		definition.NewWorkflowKey(
+			s.namespace.ID().String(),
+			workflowID,
+			workflowConditionFailed.RunID,
+		),
+		prevExecutionUpdateAction,
+		func() (workflow.Context, workflow.MutableState, error) {
+			workflowContext, err := api.NewWorkflowWithSignal(
+				ctx,
+				s.shardCtx,
+				s.namespace,
 				workflowID,
-				prevRunID,
-			),
-			prevExecutionUpdateAction,
-			func() (workflow.Context, workflow.MutableState, error) {
-				workflowContext, err := api.NewWorkflowWithSignal(
-					ctx,
-					shard,
-					namespaceEntry,
-					workflowID,
-					runID.String(),
-					startRequest,
-					nil)
-				if err != nil {
-					return nil, nil, err
-				}
-				return workflowContext.GetContext(), workflowContext.GetMutableState(), nil
-			},
-			shard,
-			workflowConsistencyChecker,
+				runID,
+				s.request,
+				nil)
+			if err != nil {
+				return nil, nil, err
+			}
+			mutableState := workflowContext.GetMutableState()
+			mutableStateInfo, err = extractMutableStateInfo(mutableState)
+			if err != nil {
+				return nil, nil, err
+			}
+			return workflowContext.GetContext(), mutableState, nil
+		},
+		s.shardCtx,
+		s.workflowConsistencyChecker,
+	)
+	return mutableStateInfo, err
+}
+
+// respondToRetriedRequest provides a response in case a start request is retried.
+func (s *Starter) respondToRetriedRequest(
+	ctx context.Context,
+	runID string,
+) (*historyservice.StartWorkflowExecutionResponse, error) {
+	request := s.request.StartRequest
+
+	if !request.GetRequestEagerExecution() {
+		return &historyservice.StartWorkflowExecutionResponse{
+			RunId: runID,
+		}, nil
+	}
+
+	// For eager workflow execution, we need to get the task info and history events in order to construct a poll response.
+	mutableStateInfo, err := s.getMutableStateInfo(ctx, runID)
+	if err != nil {
+		return nil, err
+	}
+
+	// The current workflow task is not inflight or not the first task or we exceeded the first attempt and fell back to
+	// matching based dispatch.
+	if !mutableStateInfo.hasInflight || mutableStateInfo.workflowTaskInfo.StartedEventID != 3 || mutableStateInfo.workflowTaskInfo.Attempt > 1 {
+		return nil, serviceerror.NewWorkflowExecutionAlreadyStarted(
```

Comment:
Why return error here? Start workflow is still a success.

Reply:
Because we cannot return an eager task.
Caller should be notified with an error and handle this case.
I think it's clearer than returning nil task.

Reply:
I'm on the fence whether this should result in an error or not.
In any case what the caller should do is get a handle to the workflow and wait for its completion.
With the error approach there's at least a way to let the caller know what happened.

Reply:
@Spikhalskiy @cretz I could use your opinion here

Reply:
Personally, I think it's clearer/safer to consider eagerness to be a request of the server, not a requirement. The server is allowed to use whatever heuristics it wants to deny the request and do a non-eager. The absence of a task I think is enough to tell the caller the server denied the request.

I understand your use case is a "require to be eager" but that use case isn't supported for activities either. If we want a require-to-be-eager to be a thing, it should be a thing on both. I could support a server-side/namespace option of "fail if eager requested but cannot be given".

Reply:
Just to clarify, the question here is just whether or not to respond with an error.
The behavior is going to be the same server-side.

The client may choose to use the run ID given in the error to connect to the existing workflow in both cases.
The error path has some more information that the client might find useful.

Reply:
Hrmm...k.

So if I call this start, may a workflow start because of it? If so, IMO that should never error even if something post-start can't happen. To me success means "workflow started because of this request" and failure means "workflow did not start because of this request". Unsure if that's related. Sorry, not familiar w/ details here so I don't have a big opinion.

Reply:
If a client gets this error it means that a retried request came in too late and the workflow task cannot be dispatched eagerly (likely due to it already being dispatched via the standard, matching based, path).

The only thing an error gives you over the alternative where the server responds successfully but omits the inline task is that additional piece of information.

Reply:
Simply - if success means workflow started by this request and failure means workflow not started by request, works for me. If failure can mean workflow still started by this request, that's confusing to me.

Reply:
Here the workflow was started from a previous attempt of the same request. So it seems like you're saying it should not be an error.

I tend to agree but I will add a log to avoid losing some of this information.

Reply:
Ended up returning successful result and recording a metric explaining why eager execution was denied.